In [2]:
import tensorflow as tf


Load and prepare the [MNIST](http://yann.lecun.com/exdb/mnist/) dataset.

the MNIST dataset is a large database of handwritten digits.

In [3]:
mnist = tf.keras.datasets.mnist

Convert the samples from integers to floating-point numbers

In [4]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

11493376/11490434 [==============================] - 3s 0us/step


Build the tf.keras.Sequential model by stacking layers

Sequential groups a linear stack of layers into a tf.keras.Model.

In [10]:
model = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)), # Convert Matrix to single array
    tf.keras.layers.Dense(128, activation='relu'), # first layer
    tf.keras.layers.Dropout(0.2), # randomly selected neurons are ignored
    tf.keras.layers.Dense(10) # second layer
])

In [11]:
predictions = model(x_train[:1]).numpy()
predictions

array([[ 0.09429107,  0.48994   , -0.09670943,  0.03879449, -0.09811139,
        -0.6250849 , -0.32167333, -0.17129496,  0.34362987, -0.00857204]],
      dtype=float32)

The tf.nn.softmax function converts these logits to "probabilities" for each class:

In [12]:
tf.nn.softmax(predictions).numpy()

array([[0.1089157 , 0.16177769, 0.08997877, 0.10303591, 0.08985271,
        0.05304806, 0.0718521 , 0.08351181, 0.1397581 , 0.09826925]],
      dtype=float32)

In [14]:
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
loss_fn(y_train[:1], predictions).numpy()

2.9365568

In [15]:
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

# adjusts the model parameters to minimize the loss
model.fit(x_train, y_train, epochs=5) 

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.3011 - accuracy: 0.9129
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1460 - accuracy: 0.9567
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1094 - accuracy: 0.9668
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0925 - accuracy: 0.9717
Epoch 5/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0758 - accuracy: 0.9769


In [16]:
# this method checks the models performance
model.evaluate(x_test, y_test, verbose=2)

313/313 - 0s - loss: 0.0752 - accuracy: 0.9759


[0.07521386444568634, 0.9758999943733215]

The image classifier is now trained to ~98% accuracy on this dataset